File: MetaBrain/observability_py/request_lineage.py
from .types import ObservabilityEvent, RequestLineage


def build_request_lineage(trace_id: str, events: list[ObservabilityEvent]) -> RequestLineage:
    scoped = [event for event in events if event.trace_id == trace_id]
    return RequestLineage(
        trace_id=trace_id,
        providers_used=sorted({str(event.payload_summary.get("provider_name", "unknown")) for event in scoped if event.layer == "provider"}),
        memory_accessed=any(event.layer == "memory" for event in scoped),
        imaging_used=any(event.layer == "imaging" for event in scoped),
        review_triggered=any(event.layer == "review" for event in scoped),
        confidence_generated=any(event.layer == "confidence" for event in scoped),
        escalations=[event.event_type for event in scoped if "escalation" in event.event_type],
        fallbacks=[event.event_type for event in scoped if "fallback" in event.event_type],
        blocked_outputs=[event.event_type for event in scoped if "blocked" in event.event_type],
        final_status=scoped[-1].event_type if scoped else "unknown",
    )

File: MB-Whatsapp/observability/request-lineage.ts
import type { ObservabilityEvent, RequestLineage } from "./types";

export function buildRequestLineage(trace_id: string, events: ObservabilityEvent[]): RequestLineage {
  const scoped = events.filter((event) => event.trace_id === trace_id);
  return {
    trace_id,
    providers_used: [...new Set(scoped.filter((event) => event.layer === "provider").map((event) => String(event.payload_summary.provider_name ?? "unknown")))],
    memory_accessed: scoped.some((event) => event.layer === "memory"),
    imaging_used: scoped.some((event) => event.layer === "imaging"),
    review_triggered: scoped.some((event) => event.layer === "review"),
    confidence_generated: scoped.some((event) => event.layer === "confidence"),
    escalations: scoped.filter((event) => event.event_type.includes("escalation")).map((event) => event.event_type),
    fallbacks: scoped.filter((event) => event.event_type.includes("fallback")).map((event) => event.event_type),
    blocked_outputs: scoped.filter((event) => event.event_type.includes("blocked")).map((event) => event.event_type),
    final_status: scoped.at(-1)?.event_type ?? "unknown",
  };
}
